### ISSUE TITLE
Love this!

### ISSUE THREAD

**Issue by nackjicholson** [seandstewart/yesql#6]
Looks great, man! It looks awesome, in fact. I can't fully tell what's going on yet because the docs don't show the underlying sql code. It looks like you've solved how to use type annotations in a natural way, and how to marshal the returned database data into python types.

How does it work? I'm curious, what were the “aha!” moments that led to what you have here?

@zx80 you may be interested in seeing what @seandstewart has been able to do with yesql. Love the name, by the way. I don't know if you know that `anosql` was originally based on a Clojure lib called `yesql`. We've come full circle! :D

**Comment by seandstewart** [comment]
Hey @nackjicholson !

It's been a minute (understatement)

I've been noodling on this library for some time and in fact completely pivoted on the implementation a couple times. There were a few aha moments I had:

1. Isolation of the QueryPackage from any knowledge of query execution. (I credit your library for that)
2. Isolation of query Execution to driver-specific implementations with a pre-defined API.
3. Formal Statement objects which map to the desired return-type rather than the specific query operation. (I credit [pugsql](https://github.com/mcfunley/pugsql) for that insight).
4. A better understanding of the Repository pattern (vague, but drove many of my later decisions).

Overall, I wanted to ensure that every component can operate independently. The major exceptions are Statements and Repositories, which operate as the glue which forms the public interface. I also wanted anything "magical" to be able to be overridden by the user at any point. (e.g., toggle serdes on/off, supply a custom deserializer, toggle transaction management, inject a connection, toggle auto rollbacks).

The unmarshalling itself is managed by default by my library [typical](https://github.com/seandstewart/typical) (which is in the midst of its own overhaul for v3 😅). I'm glad I could throw that over the fence though, it's a massive problem in and of itself!

Anyway, this post is getting long!

I actually found the history of these libraries in my research and that's how I ended up selecting the name! Couldn't believe it wasn't taken already.
